server/pulp/server/controllers/content.py

```diff
@@ -96,3 +110,87 @@ def refresh_content_source(content_source_id=None):
     conduit = ContentSourcesConduit('Refresh Content Source')
     step = ContentSourcesRefreshStep(conduit, content_source_id=content_source_id)
     step.process_lifecycle()
+
+
+@celery.task(base=Task)
+def _download(plugin_id, unit_locator, url, options):
+    """
+    Download the specified content unit.
+
+    :param plugin_id: An importer ID.
+    :type plugin_id: str
+    :param unit_locator: A unit locator.
+    :type unit_locator: pulp.server.db.model.UnitLocator
+    :param url: The download URL.
+    :type url: str
+    :param options: Passed to the Importer.get_downloader().
+    :type options: dict
+    """
+    _logger.info(_('Downloading: {url}').format(url=url))
+
+    # Find the plugin
+    try:
+        importer, cfg = repository.get_importer_by_id(plugin_id)
+    except PluginNotFound:
+        _logger.info(_('Download {url}, failed - plugin not found').format(url=url))
+        return
+
+    # Find the model
+    model = plugin_api.get_unit_model_by_id(unit_locator.type_id)
+    if model is None:
+        _logger.info(_('Download {url}, failed - model not found').format(url=url))
+        return
+
+    # Fetch the unit
+    unit = model.objects(id=unit_locator.unit_id).get()
+    if unit is None:
+        _logger.info(_('Download {url}, failed - unit not found').format(url=url))
+        return
+
+    if os.path.exists(unit.storage_path):
+        # Already downloaded
+        return
+
+    # Prepare the download
+    working_dir = get_working_directory()
+    destination = os.path.join(working_dir, unit_locator.unit_id)
+    request = DownloadRequest(url, destination)
+    listener = AggregatingEventListener()
+    downloader = importer.get_downloader(cfg, url, **options)
+    downloader.event_listener = listener
+
+    # Download and store the unit
+    downloader.download_one(request)
+    if listener.succeeded_reports:
+        _logger.info(_('Download {url}, succeeded').format(url=url))
+        unit.set_content(destination)
+        unit.save()
+    else:
+        _logger.info(_('Download {url}, failed').format(url=url))
```

Comment:
It might be nice to include some information about the failure from the listener report like the HTTP status code.